scheduler/job/job.go

```diff
@@ -161,54 +158,197 @@ func (j *job) Serve() {
+		if err != nil {
+			return "", err
+		}
+
+		resp.SchedulerClusterID = j.config.Manager.SchedulerClusterID
+		return internaljob.MarshalResponse(resp)
+	default:
+		msg := fmt.Sprintf("invalid scope %s", req.Scope)
+		log.Error(msg)
+		return "", fmt.Errorf("invalid scope %s", req.Scope)
+	}
+}
+
+// preheatSinglePeer preheats job by single seed peer, scheduler will trigger seed peer to download task.
+func (j *job) preheatSinglePeer(ctx context.Context, taskID string, req *internaljob.PreheatRequest, log *logger.SugaredLoggerOnWith) (*internaljob.PreheatResponse, error) {
 	// If seed peer is disabled, return error.
 	if !j.config.SeedPeer.Enable {
-		return fmt.Errorf("cluster %d scheduler %s has disabled seed peer", j.config.Manager.SchedulerClusterID, j.config.Server.AdvertiseIP)
+		return nil, fmt.Errorf("cluster %d scheduler %s has disabled seed peer", j.config.Manager.SchedulerClusterID, j.config.Server.AdvertiseIP)
 	}
 
 	// If scheduler has no available seed peer, return error.
 	if len(j.resource.SeedPeer().Client().Addrs()) == 0 {
-		return fmt.Errorf("cluster %d scheduler %s has no available seed peer", j.config.Manager.SchedulerClusterID, j.config.Server.AdvertiseIP)
-	}
-
-	req := &internaljob.PreheatRequest{}
-	if err := internaljob.UnmarshalRequest(data, req); err != nil {
-		logger.Errorf("unmarshal request err: %s, request body: %s", err.Error(), data)
-		return err
-	}
-
-	if err := validator.New().Struct(req); err != nil {
-		logger.Errorf("preheat %s validate failed: %s", req.URL, err.Error())
-		return err
+		return nil, fmt.Errorf("cluster %d scheduler %s has no available seed peer", j.config.Manager.SchedulerClusterID, j.config.Server.AdvertiseIP)
 	}
 
 	// Preheat by v2 grpc protocol. If seed peer does not support
 	// v2 protocol, preheat by v1 grpc protocol.
-	if err := j.preheatV2(ctx, req); err != nil {
-		logger.Errorf("preheat %s failed: %s", req.URL, err.Error())
-
+	resp, err := j.preheatV2(ctx, taskID, req, log)
+	if err != nil {
+		log.Errorf("preheat failed: %s", err.Error())
 		if st, ok := status.FromError(err); ok {
 			if st.Code() == codes.Unimplemented {
-				if err := j.preheatV1(ctx, req); err != nil {
+				return j.preheatV1(ctx, taskID, req, log)
+			}
+		}
+
+		return nil, err
+	}
+
+	return resp, nil
+}
+
+// preheatAllPeers preheats job by all peers, only suoported by v2 protocol. Scheduler will trigger all peers to download task.
+// If all the peer download task failed, return error. If some of the peer download task failed, return success tasks and failure tasks.
+// Notify the client that the preheat is successful.
+func (j *job) preheatAllPeers(ctx context.Context, taskID string, req *internaljob.PreheatRequest, log *logger.SugaredLoggerOnWith) (*internaljob.PreheatResponse, error) {
+	var (
+		successTasks []*internaljob.PreheatSuccessTask
+		failureTasks []*internaljob.PreheatFailureTask
+	)
+
+	sem := semaphore.NewWeighted(req.ConcurrentCount)
```

Comment:
For controlling the concurrency, can we just use the `errgroup.SetLimit(n)`?

Reply:
Fix 😊